crates/wasmparser/src/validator/func.rs

```diff
@@ -210,3 +208,567 @@ mod tests {
         assert_eq!(v.operand_stack_height(), 2);
     }
 }
+
+use super::operators::OperatorValidatorError;
+use crate::{
+    BlockType, BrTable, Ieee32, Ieee64, MemoryImmediate, SIMDLaneIndex, VisitOperator, V128,
+};
+
+impl<T> FuncValidator<T>
+where
+    T: WasmModuleResources,
+{
+    fn visit<F>(&mut self, offset: usize, visit_fn: F) -> Result<()>
+    where
+        F: FnOnce(&mut Self) -> Result<(), OperatorValidatorError>,
+    {
+        visit_fn(self)
+            .map_err(|e| e.set_offset(offset))
+            .map_err(Into::into)
+    }
+}
+
+macro_rules! forward {
+    ( $this:ident.$visit_fn:ident($offset:expr $(, $param:expr)*) $(,)? ) => {{
+        $this.visit($offset, |this| this.validator.$visit_fn(&this.resources, $( $param ),*))
+    }};
+}
+
+#[allow(unused_variables)]
+#[rustfmt::skip]
+impl<T> VisitOperator<usize> for FuncValidator<T>
+where
+    T: WasmModuleResources,
+{
+    type Output = Result<()>;
+
+    fn visit_nop(&mut self, offset: usize) -> Self::Output { forward!(self.visit_nop(offset)) }
```

Comment:
To make this a bit easier on rustc I think it might be better to not use a closure internally where each method dispatches to `visit` and then that dispatches back to a closure. Instead I think it might be better here to instead have the macro expand to what `visit` is already doing (but with fewer closures)

Reply:
Good to know that closures are heavy on `rustc`. Will do what you proposed and see if it works equally well.